### ISSUE TITLE
Conversations that happened while disconnected get received in wrong order

### ISSUE THREAD

**Issue by marga-debian** [davidgfnet/whatsapp-purple#359]
If people talk in a group while I'm disconnected, I get the messages for one side first and then the other instead of as a conversation.  This means that I get:

(20:41:04) John: Foo
(20:41:04) John: Response to Bar. Baz
(20:41:05) John: Response to WooHoo
(20:41:07) Jane: Response to Foo. Bar
(20:41:10) Jane: Response to Baz. WooHoo.

Messages should have been intermingled and the times should have had more seconds/minutes in between.

This is with the version in HEAD and resource is Android-2.31.151-443

The same thing happened to me in two different groups, with the answers from the second person arriving in one of the cases quite a while later than all the messages from the first person.

I've now built in debug mode and I'm capturing everything, but I don't think it will happen again until I disconnect and the group gets quite a lot of traffic, so it's hard to know if I'll get useful debugging information.

**Comment by wissssam** [comment]
This happend to me too .
Besides that messages received in wrong order ,some messages isn't received at all and some messages received the next day .
this happened in group chat and in chat with some contacts .
I am using the latest git version and the resource is empty .

**Comment by marga-debian** [comment]
I've been banned now, so I won't be able to debug this any further :(

**Comment by sant527** [comment]
I also got banned. How to come out.

**Comment by axel21a** [comment]
Also I was suspended, I sent them a request to rehabilitate the number, we hope goes well, otherwise I just have to buy a new sim.
or is there some other method to simulate a cell phone number?

**Comment by wissssam** [comment]
Also I've noticed that i can't receive messages that happened while disconnected  until the sender is online .

**Comment by githubtefo** [comment]
The same thing happends here. Ive got 2 groups and these only occurs in one of them, with any of its contacts.
purple-whatsapp-0.9.0-1-i686.pkg.tar.xz
Resource Android-2.12.419

**Comment by vanfanel** [comment]
It seems I'm also not receiving messages until the sender is online

**Comment by andreas-42** [comment]
I can confirm all of the above.

Going through the logs (with the verbose output of the testing branch turned on) shows that the following just happened to me, in this order:

1. I'm not online, two people have a conversation in a group I'm in.
2. The two people go offline again.
3. I go online. The logs show that WhatsApp tries to send the whole conversation to me, every single message, in the correct order. However, this doesn't work, each single message throws an Axolotl exception (InvalidKeyIdException No key state xxxxxxxx in record), and hence the pidgin frontend doesn't show anything to me.
4. The first of the two people goes online again. Now Whatsapp sends all his messages from above again. This time it works, and I can see them - but only the ones from this one person.
5. A while after that, the second of the two people goes online again, and now I receive all messages from above from him, leading to the order in the chat window that @marga-debian described. The pidgin timestamps shown in the chat window are from when the messages appear in the window, not when they were written (so for each person they're all the same time up to a second or so).

@davidgfnet, would it help you to have the log files?

**Comment by davidgfnet** [comment]
I'm aware of that. I'll fix it ASAP so don't experiment those issues any more.
About the "empty" messages, I have no idea really, I'd need traces to understand that.
Expect an update this week with more fixes! :D

**Comment by davidgfnet** [comment]
Just pushed something to the "devel" branch and I hope to merge it soon. If you guys want to play around with it :)

**Comment by githubtefo** [comment]
I compiled the "devel" branch and it is working now for the last five minutes at least.
I'll let you know if something goes wrong.
Thanks @davidgfnet !

**Comment by githubtefo** [comment]
humm, since two days ago I don't receive any new message from my chat groups. I only receive a loop of the same 10 messages every time I open pidgin. What about you?

**Comment by davidgfnet** [comment]
Any of you tested devel branch?

**Comment by andreas-42** [comment]
Yes, still get messages in wrong order sometimes as described in #391.

**Comment by giordano** [comment]
I'm using `devel` branch, updated to latest version (https://github.com/davidgfnet/whatsapp-purple/commit/5e0d80ab8c708cc425b2c08cec0fd11b41776a7e).  Unfortunately, the plugin seems somewhat broken with regards to group chats.  One group chat receives a bunch of old messages when it connects (about 10 or so), it keeps disconnecting and reconnecting every ~15 minutes, every time receiving other new ~10 messages when rejoining the chat (I've been without running Pidgin for a couple of days and I missed dozens of messages).  In another chat, I'm sure I didn't receive the latest messages that have been sent one hour ago.

**Comment by davidgfnet** [comment]
Ok, anyone can share a trace? I'm curious what the issue might be, since your comment implies it works for 1to1 conversations.

**Comment by giordano** [comment]
Actually I didn't try yet with private conversations `:-)`

**Comment by andreas-42** [comment]
@davidgfnet, do you mean a trace for messages in wrong order or for receiving multiple copies of group messages?
